chia-bls/src/signature.rs

```diff
@@ -1,44 +1,67 @@
-use crate::public_key::PublicKey;
-use crate::secret_key::SecretKey;
-use bls12_381_plus::{
-    multi_miller_loop, ExpandMsgXmd, G1Affine, G2Affine, G2Prepared, G2Projective,
-};
+use crate::{PublicKey, SecretKey};
+use blst::*;
 use chia_traits::chia_error::{Error, Result};
 use chia_traits::{read_bytes, Streamable};
-use group::{Curve, Group};
 use sha2::{Digest, Sha256};
 use std::borrow::Borrow;
 use std::convert::AsRef;
 use std::io::Cursor;
-use std::ops::Neg;
+use std::mem::MaybeUninit;
 
-#[derive(PartialEq, Eq, Debug)]
-pub struct Signature(pub(crate) G2Projective);
+// we use the augmented scheme
+pub const DST: &[u8] = b"BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_AUG_";
+
+#[derive(Debug, Clone)]
+pub struct Signature(pub(crate) blst_p2);
 
 impl Signature {
     pub fn from_bytes(buf: &[u8; 96]) -> Result<Self> {
-        match G2Affine::from_compressed(buf).into() {
-            Some(p) => Ok(Self(G2Projective::from(&p))),
-            None => Err(Error::Custom("Signature is invalid".to_string())),
+        let p2 = unsafe {
+            let mut p2_affine = MaybeUninit::<blst_p2_affine>::uninit();
+            if blst_p2_uncompress(p2_affine.as_mut_ptr(), buf as *const u8)
+                != BLST_ERROR::BLST_SUCCESS
+            {
+                return Err(Error::Custom("Signature is invalid".to_string()));
+            }
+            let mut p2 = MaybeUninit::<blst_p2>::uninit();
+            blst_p2_from_affine(p2.as_mut_ptr(), &p2_affine.assume_init());
+            p2.assume_init()
+        };
+        let ret = Self(p2);
+        if !ret.is_valid() {
+            Err(Error::Custom("Signature is invalid".to_string()))
+        } else {
+            Ok(ret)
         }
     }
 
     pub fn to_bytes(&self) -> [u8; 96] {
-        self.0.to_affine().to_compressed()
+        unsafe {
+            let mut bytes = MaybeUninit::<[u8; 96]>::uninit();
+            blst_p2_compress(bytes.as_mut_ptr() as *mut u8, &self.0);
+            bytes.assume_init()
+        }
     }
 
     pub fn aggregate(&mut self, sig: &Signature) {
-        self.0 += sig.0;
+        unsafe {
+            blst_p2_add(&mut self.0, &self.0, &sig.0);
```

Comment:
in bls-signature we used `blst_p2_add_or_double`

Reply:
oh, I've fixed this locally, but in a later commit, that I didn't include here. I'll fix it.

Reply:
I added a commit to address this, including a test case that fails without the fix